### PR TITLE
Case 21228: [Create App] allow editing properties of multiple entities simultaneously

### PR DESCRIPTION
**Pull request by thoys**
https://highfidelity.manuscript.com/f/cases/21228
https://highfidelity.manuscript.com/f/cases/19989

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18664/HighFidelity-Beta-PR15360-aa6a374.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18664/HighFidelity-Beta-Interface-PR15360-aa6a374.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18664/HighFidelity-Beta-PR15360-aa6a374.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18664/HighFidelity-Quest-Beta-PR15360-aa6a374.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18664/HighFidelity-Beta-PR15360-aa6a374.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18664/HighFidelity-Beta-Interface-PR15360-aa6a374.exe

File: scripts/system/html/js/entityProperties.js [hunk 1]
```diff
@@ -3279,20 +3568,369 @@ function materialTargetPropertyUpdate(propertyValue) {
...
+                    property.elInput.classList.toggle('multi-diff');
+                    if (isMultiDiffValue) {
+                        property.elInput.setMultipleValues();
+                    } else {
+                        property.elInput.imageLoad(property.elInput.value);
+                    }
+                    break;
+                }
+                case 'dynamic-multiselect': {
+                    if (!isMultiDiffValue && property.data.propertyUpdate) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: I think we need to disable dynamic multiselects for multi edit

File: scripts/system/html/js/draggableNumber.js [hunk 2]
```diff
@@ -70,22 +72,26 @@ DraggableNumber.prototype = {
...
-                        this.elInput.stepUp();
-                        --dx;
-                    } else {
-                        this.elInput.stepDown();
-                        ++dx;
+            let changeDelta = event.clientX - this.lastMouseEvent.clientX;
+            if (changeDelta !== 0) {
+                if (this.multiDiffModeEnabled) {
+                    if (this.multiDiffDragFunction) {
+                        this.multiDiffDragFunction(changeDelta * this.step);
```
Comment (dback2) [inline scripts/system/html/js/draggableNumber.js]: I think stepUp and stepDown need to work the same as drag for multi diff in terms of it being a delta change. Right now it sets them all to the same value then steps.

File: scripts/system/html/js/entityProperties.js [hunk 3]
```diff
@@ -1772,43 +1860,173 @@ function resetProperties() {
...
+
+    return {
+        value: propertyValues[0],
+        values: propertyValues,
+        isMultiDiffValue: false
+    };
+}
+
+
+function getDetailedNumberMPVDiff(multiplePropertyValue, propertyData) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Feels like this could maybe have a better function name

File: scripts/system/html/js/entityProperties.js [hunk 4]
```diff
@@ -1772,43 +1860,173 @@ function resetProperties() {
...
+            }
+            detailedValues[DEFAULT_KEY].push(applyInputNumberPropertyModifiers(propertyValue, propertyData));
+            unmodifiedValues[DEFAULT_KEY].push(propertyValue);
         }
-    } else {
-        propertyValue = selectedEntityProperties[originalPropertyName];
+    });
+    let keys = [...uniqueKeys];
+
+    let subPropertyDiff = {};
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Using subProperty naming here for subPropertyDiff and averagePerSubProperty could be confusing because subProperty has otherwise referred to the collidesWith sub property elements.

File: scripts/system/html/js/entityProperties.js [hunk 5]
```diff
@@ -1772,43 +1860,173 @@ function resetProperties() {
...
+    });
+
+    return {
+        keys,
+        subPropertyDiff,
+        averagePerSubProperty
+    };
+}
+
+function getDetailedSubPropertyMVPDiff(multiplePropertyValue, subPropertyName) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Feels like this could maybe have a better function name (also isn't consistent with MPV naming above)

File: scripts/system/html/js/entityProperties.js [hunk 6]
```diff
@@ -1903,60 +2138,71 @@ function createDragStartFunction(property) {
...
-            z: property.elNumberZ.elInput.value * multiplier
-        };
-        updateProperty(property.name, newValue, property.isParticleProperty);
+        let value = parseFloat(applyOutputNumberPropertyModifiers(parseFloat(this.value), property.data));
+        updateProperty(property.name, value, property.isParticleProperty);
     };
 }
 
-function createEmitRectPropertyUpdateFunction(property) {
+function createEmitNumberSubPropertyUpdateFunction(property, subProperty) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: This is not a subProperty like how collidesWith works, so we should have separate naming for these different "subProperty" cases.

File: scripts/system/html/js/entityProperties.js [hunk 7]
```diff
@@ -1903,60 +2138,71 @@ function createDragStartFunction(property) {
...
+                let updateObject = createPropertyUpdateObject(property.name, propertyObject);
+                updateObjects.push({
+                    entityIDs: [entityID],
+                    properties: updateObject,
+                });
+
+                mergeDeep(currentSelections[i].properties, updateObject);
+            }
+
+            // only update the entity property value itself if in the middle of dragging
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: This onlyUpdateEntity check is now in 2 places so maybe should put in a function

File: scripts/system/html/js/entityProperties.js [hunk 8]
```diff
@@ -1903,60 +2138,71 @@ function createDragStartFunction(property) {
...
+            // prevent undo command push, saving new property values, and property update
+            // callback until drag is complete (additional update sent via dragEnd callback)
+            let onlyUpdateEntity = properties[property.name] && properties[property.name].dragging === true;
+            updateMultiDiffProperties(updateObjects, onlyUpdateEntity);
+            console.log('updateMultiDiffProperties');
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Do we need this?

File: scripts/system/html/js/entityProperties.js [hunk 9]
```diff
@@ -2101,24 +2376,75 @@ function updateNumberMinMax(property) {
...
 }
 
+/**
+ *
+ * @param {object} property - property update on drag
+ * @param {string} [subProperty] - subProperty to update on drag (e.g. enter 'x' to just update position.x)
+ * @returns {Function}
+ */
+function createMultiDiffDragFunction(property, subProperty) {
+    return function(changedDelta) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: as in draggableNumber.js, changedDelta is sorta redundant so something like deltaX would be better

File: scripts/system/html/js/entityProperties.js [hunk 10]
```diff
@@ -2101,24 +2376,75 @@ function updateNumberMinMax(property) {
...
+            return;
+        }
+        let updateObjects = [];
+        const selectedEntityIDsArray = [...selectedEntityIDs];
+
+        for (let i = 0; i < selectedEntityIDsArray.length; ++i) {
+            let entityID = selectedEntityIDsArray[i];
+
+            let updatedValue;
+            if (subProperty !== undefined) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: as mentioned elsewhere, subProperty naming here conflicts with the sense of subProperties for collidesWith, I would maybe use propertyVar instead

File: scripts/system/html/js/entityProperties.js [hunk 11]
```diff
@@ -3279,20 +3568,369 @@ function materialTargetPropertyUpdate(propertyValue) {
...
+                    break;
+                }
+                case 'icon': {
+                    property.elSpan.innerHTML = propertyData.icons[propertyValue];
+                    property.elSpan.style.display = "inline-block";
+                    break;
+                }
+                case 'texture': {
+                    property.elInput.value = isMultiDiffValue ? "" : propertyValue;
+                    property.elInput.classList.toggle('multi-diff');
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Should this have a value with the toggle?

File: scripts/system/html/js/entityProperties.js [hunk 12]
```diff
@@ -3279,20 +3568,369 @@ function materialTargetPropertyUpdate(propertyValue) {
...
+                    let detailedNumberDiff = getDetailedNumberMPVDiff(propertyMultiValue, propertyData);
+                    property.elNumberX.setValue(detailedNumberDiff.averagePerSubProperty.x, detailedNumberDiff.subPropertyDiff.x);
+                    property.elNumberY.setValue(detailedNumberDiff.averagePerSubProperty.y, detailedNumberDiff.subPropertyDiff.y);
+                    if (property.elNumberZ !== undefined) {
+                        property.elNumberZ.setValue(detailedNumberDiff.averagePerSubProperty.z, detailedNumberDiff.subPropertyDiff.z);
+                    }
+                    break;
+                }
+                case 'color': {
+                    let displayColor = propertyMultiValue.isMultiDiffValue ? propertyMultiValue.values[0] : propertyValue;
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: The design was to use the first color value for multivalues here?

File: scripts/system/html/js/entityProperties.js [hunk 13]
```diff
@@ -3279,20 +3568,369 @@ function materialTargetPropertyUpdate(propertyValue) {
...
         }
         elInput.checked = materialTargets.indexOf(materialTargetName) >= 0;
     }
     
     elDivOptions.propertyValue = propertyValue;
 }
 
+function roundAndFixNumber(number, propertyData) {
+    let result = number;
+    if (propertyData.round !== undefined) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: You are calling roundAndFixNumber from applyOutputNumberPropertyModifiers from createEmitNumberPropertyUpdateFunction, but previously rounding only occurred when a draggable number or vector property value update was received.

File: scripts/system/html/js/entityProperties.js [hunk 14]
```diff
@@ -3279,20 +3568,369 @@ function materialTargetPropertyUpdate(propertyValue) {
...
+        result = Math.round(result * propertyData.round) / propertyData.round;
+    }
+    if (propertyData.decimals !== undefined) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Decimals adjustments also only previously occurred for draggable numbersor vector property value updates being set.

File: scripts/system/html/js/entityProperties.js [hunk 15]
```diff
@@ -3279,20 +3568,369 @@ function materialTargetPropertyUpdate(propertyValue) {
...
+
+const isSetsEqual = (a, b) => a.size === b.size && [...a].every(value => b.has(value));
+
+
+function handleEntitySelectionUpdate(selections, isPropertiesToolUpdate) {
+    const previouslySelectedEntityIDs = selectedEntityIDs;
+    currentSelections = selections;
+    selectedEntityIDs = new Set(selections.map(selection => selection.id));
+    const multipleSelections = currentSelections.length > 1;
+    const hasSelectedEntityChanged = !isSetsEqual(selectedEntityIDs, previouslySelectedEntityIDs);
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: areSetsEqual

File: scripts/system/html/js/entityProperties.js [hunk 16]
```diff
@@ -3279,20 +3568,369 @@ function materialTargetPropertyUpdate(propertyValue) {
...
+        showNewJSONMaterialEditorButton();
+
+        disableProperties();
+    } else {
+        if (!isPropertiesToolUpdate && !hasSelectedEntityChanged && document.hasFocus()) {
+            // in case the selection has not changed and we still have focus on the properties page,
+            // we will ignore the event.
+            return;
+        }
+        if (hasSelectedEntityChanged) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: This should maybe be changed into a system that supports enabled/disabling multi properties for multi edit or not as parentMaterialName / Material Target property should fall under this too (technically all dynamic-multiselects probably should even though there is only one right now).

File: scripts/system/html/js/entityProperties.js [hunk 17]
```diff
@@ -3123,14 +3412,14 @@ function toggleDropdown(event) {
...
-    elDropdowns = document.querySelectorAll("div.dropdown > dl");
+    let elDropdowns = document.querySelectorAll("div.dropdown > dl");
     for (let i = 0; i < elDropdowns.length; ++i) {
         elDropdowns[i].setAttribute('dropped', 'false');
     }
 }
 
 function setDropdownValue(event) {
-    let dt = event.target.parentNode.parentNode.previousSibling;
+    let dt = event.target.parentNode.parentNode.previousSibling.previousSibling;
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: since there is a new `multi-diff` span element under the dropdown now, which would otherwise be selected instead.
image

File: scripts/system/html/css/edit-style.css [hunk 18]
```diff
@@ -333,7 +367,7 @@ input[type=range]::-webkit-slider-thumb {
 input[type=range]::-webkit-slider-thumb:hover {
     background-color: white;
 }
-input[type=range]:focus { /*#252525*/
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/html/css/edit-style.css]: 🤔

File: scripts/system/html/js/draggableNumber.js [hunk 19]
```diff
@@ -124,7 +130,15 @@ DraggableNumber.prototype = {
...
     },
     
-    setValue: function(newValue) {
+    setValue: function(newValue, isMultiDiff) {
+        if (isMultiDiff !== undefined) {
+            this.setMultiDiff(isMultiDiff);
+        }
+
+        if (isNaN(newValue)) {
+            throw newValue + " is not a number";
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/html/js/draggableNumber.js]: Since we want to know that something is broken and where it is broken instead of simply ignoring and returning. But if you feel strongly about this we can discuss it.

File: scripts/system/html/js/entityProperties.js [hunk 20]
```diff
@@ -1585,11 +1595,80 @@ let properties = {};
...
-let selectedEntityProperties;
-let lastEntityID = null;
+let selectedEntityIDs = new Set();
+let currentSelections = [];
 let createAppTooltip = new CreateAppTooltip();
 let currentSpaceMode = PROPERTY_SPACE_MODE.LOCAL;
 
+
+
+// mergeDeep function from https://stackoverflow.com/a/34749873
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: these utils are es6 only, so that would be a utils file for scripts that run on the web-engine only, but sure.

File: scripts/system/html/js/entityProperties.js (hunk 11 above)
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: good find

File: scripts/system/html/js/entityProperties.js (hunk 13 above)
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: Correct, but numbers that do slightly differ behind their rounding shouldn't be multi-diff. Eg. if you round 2 digits and the 2 values are `4.0202` and `4.0205` I prefer it to not see that as a different since they both will be displayed as `4.02`  if that makes sense

File: scripts/system/html/js/entityProperties.js (hunk 12 above)
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: yes

File: scripts/system/edit.js [hunk 21]
```diff
@@ -2361,41 +2361,64 @@ var PropertiesTool = function (opts) {
...
-                if (data.properties.localRotation !== undefined) {
-                    data.properties.localRotation = Quat.fromVec3Degrees(data.properties.localRotation);
-                }
-                if (data.properties.emitOrientation !== undefined) {
-                    data.properties.emitOrientation = Quat.fromVec3Degrees(data.properties.emitOrientation);
+
+            if (data.properties || data.propertiesMap) {
+                var propertiesMap = data.propertiesMap;
+                if (propertiesMap === undefined) {
+                    var updateEntityIDs = data.ids !== undefined ? data.ids : selectionManager.selections;
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/edit.js]: Fail safe, I thought that entityList used the updates too for renaming/updating locks, but that didn't seem to be the case.

File: scripts/system/html/js/entityProperties.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: done

File: scripts/system/edit.js [hunk 22]
```diff
@@ -2405,9 +2428,9 @@ var PropertiesTool = function (opts) {
             selectionManager._update(false, this);
             blockPropertyUpdates = false;
         } else if (data.type === 'saveUserData' || data.type === 'saveMaterialData') {
-            //the event bridge and json parsing handle our avatar id string differently.
-            var actualID = data.id.split('"')[1];
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/edit.js]: Correct, it works fine without all that.

File: scripts/system/html/js/draggableNumber.js [hunk 23]
```diff
@@ -70,22 +72,26 @@ DraggableNumber.prototype = {
...
-            let changeValue = dx !== 0;
-            if (changeValue) {
-                while (dx !== 0) {
-                    if (dx > 0) {
-                        this.elInput.stepUp();
-                        --dx;
-                    } else {
-                        this.elInput.stepDown();
-                        ++dx;
+            let changeDelta = event.clientX - this.lastMouseEvent.clientX;
```
[the remark this answers is not in the document]
Comment (thoys) [reply scripts/system/html/js/draggableNumber.js]: dragDelta it is

File: scripts/system/html/js/entityProperties.js (hunk 9 above)
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: dragDelta it is

File: scripts/system/html/js/entityProperties.js (hunk 8 above)
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: Nop, it is gone.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18705/HighFidelity-Beta-PR15360-00a024f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18705/HighFidelity-Beta-Interface-PR15360-00a024f.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18705/HighFidelity-Beta-PR15360-00a024f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18705/HighFidelity-Quest-Beta-PR15360-00a024f.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18705/HighFidelity-Beta-PR15360-00a024f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18705/HighFidelity-Beta-Interface-PR15360-00a024f.exe

File: scripts/system/edit.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (dback2) [reply scripts/system/edit.js]: Would it maybe be better to try to catch cases that come in that way vs just using selectionManager selections? Could see the possibility of a bug later on.

File: scripts/system/html/js/draggableNumber.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (dback2) [reply scripts/system/html/js/draggableNumber.js]: I am normally of the opinion that things should fail gracefully, we can maybe log instead and not halt execution.

File: scripts/system/html/js/entityProperties.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (dback2) [reply scripts/system/html/js/entityProperties.js]: There is a utils file for web-engine side already.

File: scripts/system/html/js/entityProperties.js [hunk 24]
```diff
@@ -1775,43 +1878,180 @@ function resetProperties() {
...
-        propertyValue = selectedEntityProperties[originalPropertyName];
+    });
+    let keys = [...uniqueKeys];
+
+    let childPropertyDiff = {};
+    Object.entries(detailedValues).forEach(function([key, value]) {
+        childPropertyDiff[key] = [...new Set(value)].length > 1;
+    });
+
+    let averagePerSubProperty = {};
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: averagePerChildProperty

File: scripts/system/html/js/entityProperties.js (hunk 4 above)
Comment (dback2) [reply scripts/system/html/js/entityProperties.js]: I know we agreed on childProperty earlier but thinking about it more, it is still almost as confusing as having both as subProperty, because it is still making it seem like it is a property when it isn't. I think we need to go with propertyVar/Variable or something more like that.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18725/HighFidelity-Beta-PR15360-ef7c34e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18725/HighFidelity-Beta-Interface-PR15360-ef7c34e.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18725/HighFidelity-Beta-PR15360-ef7c34e.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18725/HighFidelity-Quest-Beta-PR15360-ef7c34e.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18725/HighFidelity-Beta-PR15360-ef7c34e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18725/HighFidelity-Beta-Interface-PR15360-ef7c34e.exe

File: scripts/system/html/js/entityProperties.js [hunk 25]
```diff
@@ -1775,43 +1811,194 @@ function resetProperties() {
...
+    });
+
+    return {
+        keys,
+        propertyComponentDiff,
+        averagePerPropertyComponent,
+    };
+}
+
+function getDetailedSubPropertyMPVDiff(multiplePropertyValue, subPropertyName) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Still named subProperty

File: scripts/system/html/js/entityProperties.js [hunk 26]
```diff
@@ -3643,16 +3653,6 @@ function handleEntitySelectionUpdate(selections, isPropertiesToolUpdate) {
     const multipleSelections = currentSelections.length > 1;
     const hasSelectedEntityChanged = !areSetsEqual(selectedEntityIDs, previouslySelectedEntityIDs);
 
-    // FIXME: do we really want to save userData/materialData here instead of saving it on the blur event of the json editor?
-    if (hasSelectedEntityChanged) {
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Why is this removed?

File: scripts/system/html/js/entityProperties.js [hunk 27]
```diff
@@ -3643,16 +3653,6 @@ function handleEntitySelectionUpdate(selections, isPropertiesToolUpdate) {
     const multipleSelections = currentSelections.length > 1;
     const hasSelectedEntityChanged = !areSetsEqual(selectedEntityIDs, previouslySelectedEntityIDs);
 
-    // FIXME: do we really want to save userData/materialData here instead of saving it on the blur event of the json editor?
```
Comment (dback2) [inline scripts/system/html/js/entityProperties.js]: Why is this removed?

File: scripts/system/html/js/draggableNumber.js [hunk 28]
```diff
@@ -70,22 +72,22 @@ DraggableNumber.prototype = {
...
-                        --dx;
+            let dragDelta = event.clientX - this.lastMouseEvent.clientX;
+            if (dragDelta !== 0) {
+                if (this.multiDiffModeEnabled) {
+                    if (this.multiDiffDragFunction) {
+                        this.multiDiffDragFunction(dragDelta * this.step);
+                    }
+                } else {
+                    if (dragDelta > 0) {
+                        this.elInput.stepUp(dragDelta);
```
Comment (dback2) [inline scripts/system/html/js/draggableNumber.js]: stepUp and stepDown still aren't working as a dragDelta they set a hard value on first step

File: scripts/system/html/js/entityProperties.js (hunk 27 above)
Comment (thoys) [reply scripts/system/html/js/entityProperties.js]: It caused problems when switching between material entities the material data actively changed. I assume the same for userData since that basically is the same kind of field. If you want to save user/material data, click save. Doing it on switch is a bit tricky.

File: scripts/system/html/js/draggableNumber.js (hunk 28 above)
Comment (thoys) [reply scripts/system/html/js/draggableNumber.js]: How do you mean? documentMouseMove should be called continuous when dragging, every time it is called the dragDelta will contain the drag delta(difference) between those frames

Comment (dback2) [reply scripts/system/html/js/draggableNumber.js]: Sorry I meant the act of stepping up and stepping down are not doing a dragDelta and set a hard value when using the arrows.

Comment (dback2) [reply scripts/system/html/js/draggableNumber.js]: This should have been on line 79 my bad hehe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18791/HighFidelity-Beta-PR15360-dc9f889.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18791/HighFidelity-Beta-Interface-PR15360-dc9f889.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18791/HighFidelity-Beta-PR15360-dc9f889.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18791/HighFidelity-Quest-Beta-PR15360-dc9f889.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18791/HighFidelity-Beta-PR15360-dc9f889.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18791/HighFidelity-Beta-Interface-PR15360-dc9f889.exe

File: scripts/system/html/js/draggableNumber.js (hunk 28 above)
Comment (thoys) [reply scripts/system/html/js/draggableNumber.js]: Thanks for clarifying again, I have provided a fix in the latest commit

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18793/HighFidelity-Beta-PR15360-386d392.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18793/HighFidelity-Beta-Interface-PR15360-386d392.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18793/HighFidelity-Beta-PR15360-386d392.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18793/HighFidelity-Quest-Beta-PR15360-386d392.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18793/HighFidelity-Beta-PR15360-386d392.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18793/HighFidelity-Beta-Interface-PR15360-386d392.exe
